src/python/grpcio_observability/grpc_observability/client_call_tracer.cc

```diff
@@ -0,0 +1,242 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "src/python/grpcio_observability/grpc_observability/client_call_tracer.h"
+
+#include "absl/strings/escaping.h"
+#include "absl/strings/string_view.h"
+
+#include <limits.h>
+
+#include <atomic>
+
+
+namespace grpc_observability {
+
+constexpr uint32_t
+    PythonOpenCensusCallTracer::PythonOpenCensusCallAttemptTracer::kMaxTraceContextLen;
+constexpr uint32_t
+    PythonOpenCensusCallTracer::PythonOpenCensusCallAttemptTracer::kMaxTagsLen;
+
+//
+// OpenCensusCallTracer
+//
+
+PythonOpenCensusCallTracer::PythonOpenCensusCallTracer(char* method, char* trace_id,
+                                                       char* parent_span_id, bool tracing_enabled):
+      method_(GetMethod(method)),
+      tracing_enabled_(tracing_enabled) {
+  GenerateClientContext(absl::StrCat("Sent.", method_), absl::string_view(trace_id),
+                        absl::string_view(parent_span_id), &context_);
+}
+
+void PythonOpenCensusCallTracer::GenerateContext() {}
+
+void PythonOpenCensusCallTracer::RecordAnnotation(absl::string_view annotation) {
+  // If tracing is disabled, the following will be a no-op.
+  context_.AddSpanAnnotation(annotation);
+}
+
+
+PythonOpenCensusCallTracer::~PythonOpenCensusCallTracer() {
+  if (PythonOpenCensusStatsEnabled()) {
+    std::vector<Label> labels = context_.Labels();
+    labels.emplace_back(Label{kClientMethod, std::string(method_)});
+    RecordIntMetric(kRpcClientRetriesPerCallMeasureName, retries_ - 1, labels); // exclude first attempt
+    RecordIntMetric(kRpcClientTransparentRetriesPerCallMeasureName, transparent_retries_, labels);
+    RecordDoubleMetric(kRpcClientRetryDelayPerCallMeasureName, ToDoubleMilliseconds(retry_delay_), labels);
+  }
+
+  if (tracing_enabled_) {
+    context_.EndSpan();
+    if (IsSampled()) {
+      RecordSpan(context_.Span().ToCensusData());
+    }
+  }
+}
+
+
+PythonCensusContext PythonOpenCensusCallTracer::CreateCensusContextForCallAttempt() {
+  auto context = PythonCensusContext(absl::StrCat("Attempt.", method_), &(context_.Span()), context_.Labels());
```

Comment:
Why "." as a separator between "Attempt" and the method? Seems like " " might be a better delimiter since " " is not a legal character in method names.